### DISCUSSION TITLE
spack install and exodus.py

### DISCUSSION THREAD

**Discussion by aperijake** [sandialabs/seacas#450] [Q&A] [answered]
Hello. I have installed SEACAS via spack on an Ubuntu 20.04 system and have been using many of the tools successfully. I recently wanted to create an Exodus file using `exodus.py`, but I can't find the script. I have tried searching for exodus.py in my spack directory, and it doesn't appear to be there.  I installed SEACAS with all the defaults (i.e., simply used `spack add seacas`). Do I need to specify something else to get `exodus.py`?

**Comment by gdsjaar** [comment]
It should build and install exodus.py and exomerge.py automatically

```
08:43 $ spack load seacas
==> Error: seacas matches multiple packages.
  Matching packages:
    kg63w7i ***@***.******@***.******@***.******@***.***> arch=darwin-sonoma-m2
    pl4xiwk ***@***.******@***.******@***.******@***.***> arch=darwin-sonoma-m2
  Use a more specific spec (e.g., prepend '/' to the hash).
08:43 $ spack load /kg63w7i
08:43 $ which blot
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt/bin/blot
08:43 $ ls /Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt/
bin     include lib
08:43 $ find /Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt/ -name \*.py
…
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt//lib/exodus3.py
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt//lib/exomerge3.py
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt//lib/exodus.py
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt//lib/exomerge.py
```

My `PYTHONPATH` also seems to be set correctly by the `spack load seacas` command:
```
08:55 $ echo $PYTHONPATH
/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt/lib:/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/xcb-proto-1.16.0-d72dp4vfokxjmvyv3dzadrqb6waigtio/lib64/python3.12/site-packages:/Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/xcb-proto-1.16.0-d72dp4vfokxjmvyv3dzadrqb6waigtio/lib/python3.12/site-packages
```
This is just the result of a `spack install seacas~mpi` although it should be the same for a `spack install seacas`
..Greg
From: Jake Koester ***@***.***>
Date: Monday, April 8, 2024 at 10:32 AM
To: sandialabs/seacas ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] [sandialabs/seacas] spack install and exodus.py (Discussion #450)

Hello. I have installed SEACAS via spack on an Ubuntu 20.04 system and have been using many of the tools successfully. I recently wanted to create an Exodus file using exodus.py, but I can't find the script. I have tried searching for exodus.py in my spack directory, and it doesn't appear to be there. I installed SEACAS with all the defaults (i.e., simply used spack add seacas). Do I need to specify something else to get exodus.py?

**Reply by aperijake** [reply]
Thanks for the help, Greg. I tried a fresh install in a new spack environment using `spack add --install seacas ~mpi`, but still don't see exodus.py.  I also have it installed on my Mac and it shows up there.

Ubuntu output:
```
azureuser@AperiAzureGPU1:~$ spack load seacas
==> Error: seacas matches multiple packages.
  Matching packages:
    ufsrfux seacas@2024-03-11%gcc@10.5.0 arch=linux-ubuntu20.04-zen2
    qwvs6ck seacas@2024-03-11%gcc@10.5.0 arch=linux-ubuntu20.04-zen2
  Use a more specific spec (e.g., prepend '/' to the hash).

azureuser@AperiAzureGPU1:~$ spack load /ufsrfux

azureuser@AperiAzureGPU1:~$ which blot
/home/azureuser/projects/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/seacas-2024-03-11-ufsrfux4lr2nhj4qmxi2scl2bi4weagd/bin/blot

azureuser@AperiAzureGPU1:~$ find /home/azureuser/projects/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/seacas-2024-03-11-ufsrfux4lr2nhj4qmxi2scl2bi4weagd/lib/ -name \*.py
azureuser@AperiAzureGPU1:~$ 
```
I tried searching the spack directory for `exodus.py`, `exodus2.py`, and `exodus3.py` and found nothing.

Mac output:
```
(base) [develop][~/projects/spack]$ spacktivate seacas                                                                 
(base) [develop][~/projects/spack]$ spack find seacas 
==> In environment seacas
==> Root specs
seacas 

==> Installed packages
-- darwin-sonoma-m1 / apple-clang@15.0.0 ------------------------
seacas@2023-10-24
==> 1 installed package
(base) [develop][~/projects/spack]$ which blot                                                                         
/Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/bin/blot
(base) [develop][~/projects/spack]$ ll /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus*py
lrwxr-xr-x  1 jake  staff   137B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus.py
lrwxr-xr-x  1 jake  staff   138B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus2.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus2.py
lrwxr-xr-x  1 jake  staff   138B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus3.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus3.py
```

**Comment by gdsjaar** [comment]
There should be a file `spack-build-out.txt` that contains the configuration and build information.  Mine is compressed and is at   520  ls /Users/gdsjaar/src/spack.fork/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/seacas-2024-04-03-kg63w7iuldibj2v435e2fxbabg3ofejt/.spack/spack-build-out.txt.gz.   Do you have one of those and can you send it to me…

Doesn’t have lots of information re python, but might be helpful…
..Greg

From: Jake Koester ***@***.***>
Date: Tuesday, April 9, 2024 at 10:54 AM
To: sandialabs/seacas ***@***.***>
Cc: Sjaardema, Greg ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [sandialabs/seacas] spack install and exodus.py (Discussion #450)

Thanks for the help, Greg. I tried a fresh install in a new spack environment using spack add --install seacas ~mpi, but still don't see exodus.py. I also have it installed on my Mac and it shows up there.

Ubuntu output:

***@***.***:~$ spack load seacas

==> Error: seacas matches multiple packages.

  Matching packages:

    ufsrfux ***@***.******@***.*** arch=linux-ubuntu20.04-zen2

    qwvs6ck ***@***.******@***.*** arch=linux-ubuntu20.04-zen2

  Use a more specific spec (e.g., prepend '/' to the hash).



***@***.***:~$ spack load /ufsrfux



***@***.***:~$ which blot

/home/azureuser/projects/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/seacas-2024-03-11-ufsrfux4lr2nhj4qmxi2scl2bi4weagd/bin/blot



***@***.***:~$ find /home/azureuser/projects/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/seacas-2024-03-11-ufsrfux4lr2nhj4qmxi2scl2bi4weagd/lib/ -name \*.py

***@***.***:~$

I tried searching the spack directory for exodus.py, exodus2.py, and exodus3.py and found nothing.

Mac output:

(base) [develop][~/projects/spack]$ spacktivate seacas

(base) [develop][~/projects/spack]$ spack find seacas

==> In environment seacas

==> Root specs

seacas



==> Installed packages

-- darwin-sonoma-m1 / ***@***.*** ------------------------

***@***.***

==> 1 installed package

(base) [develop][~/projects/spack]$ which blot

/Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/bin/blot

(base) [develop][~/projects/spack]$ ll /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus*py

lrwxr-xr-x  1 jake  staff   137B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus.py

lrwxr-xr-x  1 jake  staff   138B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus2.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus2.py

lrwxr-xr-x  1 jake  staff   138B Jan 16 16:15 /Users/jake/projects/spack/var/spack/environments/seacas/.spack-env/view/lib/exodus3.py -> /Users/jake/projects/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/seacas-2023-10-24-423wph62thxpet3q2hfhz6gl3r2cofzq/lib/exodus3.py

**Comment by aperijake** [comment]
Here it is:
[spack-build-out.txt.gz](https://github.com/sandialabs/seacas/files/14922075/spack-build-out.txt.gz)

**Comment by gdsjaar** [comment]
There is probably something missing in my seacas/package.py to put an explicit dependency on python.  Currently, it just looks to see if it can find a python executable and if so, then it installs exodus.py an exomerge.py.  In your case, it could not find a python, so it skips installing them:

-- Could NOT find Python (missing: Python_EXECUTABLE Python_INCLUDE_DIRS Python_LIBRARIES Interpreter Development Development.Module Development.Embed)
-- Exodus.py and exomerge.py NOT installed since Python could not be found.

**Reply by aperijake** [reply]
Yeah, I was just noticing the missing python. There was just `python3` and no `python`.  I added the symbolic link via:
`sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 10`
and reinstalled seacas and it seems to find python now:
[spack-build-out.txt](https://github.com/sandialabs/seacas/files/14922366/spack-build-out.txt)
I still can't find exodus.py, though.

**Reply by aperijake** [reply]
Well, I guess it finds it in one spot, but later there is:
```
-- Could NOT find Python (missing: Python_EXECUTABLE Python_INCLUDE_DIRS Python_LIBRARIES Interpreter Development Development.Module Development.Embed) 
-- Exodus.py and exomerge.py NOT installed since Python could not be found.
```

**Reply by aperijake** [reply]
Okay did:
`sudo apt install python3.8-dev`
and reinstalled, and it is there now! Thanks for the help, Greg.

**Comment by gsjaardema** [comment]
That shouldn't be needed, I'm not sure why it is looking for Python_LIBRARIES and a Development enviornment, so I will take a look at that, but glad to hear that it is now  working for you.

**Comment by gsjaardema** [comment]
I guess it is explicitly asking for Python `Development` in the scripts subdirectory.  I should be able to remove that and maybe it will then just look for the interpreter...

**Reply by gsjaardema** [reply]
This change has been done; python should now only need the interpreter and not require the development environment.  This is available in seacas@master

**Reply by aperijake** [reply, marked as answer]
Great! Thanks Greg.

So, in summary:

- My issue was I just had `python3` in my path and not `python`. 
- It required Python `Development`, but that is not really needed, and the requirement has been removed.

**Comment by gsjaardema** [comment]
> Great! Thanks Greg.
> 
> So, in summary:
> 
> * My issue was I just had `python3` in my path and not `python`.
> * It required Python `Development`, but that is not really needed, and the requirement has been removed.

Yes.  I will take a look and see how to make it work with python3 as the executable name instead of python.  I think CMake has that option, I just haven't had a chance to look (I had the same issue on my mac with python3 vs python awhile ago, but also did the symlink approach and then promptly forgot about it).
